state/protocol/badger/mutator.go

```diff
@@ -738,14 +760,25 @@ func (m *FollowerState) handleServiceEvents(block *flow.Block) ([]func(*badger.T
 	// keep track of DB operations to apply when inserting this block
 	var ops []func(*badger.Txn) error
 
+	// we will apply service events from blocks which are sealed by this block's PARENT
+	parent, err := m.blocks.ByID(block.Header.ParentID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get parent (id=%x): %w", block.Header.ParentID, err)
+	}
+
```

Comment:
1. since we are handling events for its parent, what's about passing the parent to this function instead of letting this function to query a parent?
2. ⚠️  we might handle events for the same parent again, when inserting a second child of that parent, would that raise error? Are we able to exit earlier?

Reply:
>1 since we are handling events for its parent, what about passing the parent to this function

Would this be for performance? The parent is effectively guaranteed to be cached , I don't think there would be much performance impact by passing it as an argument. We also don't already have the `parent` at the callsite
https://github.com/onflow/flow-go/blob/9c91dc26cded866644c114840d7b7241bb21a446/state/protocol/badger/mutator.go#L469-L485

> 2 we might handle events for the same parent again when inserting a second child of that parent

Good point 🙇 . For a sealing chain like `A <- B(RA) <- C(SA) <- D`, when inserting conflicting children of `C` we always need to insert the epoch status for the child `D`. This will not cause conflicts because it is keyed by the block ID of `D`. We also need to insert the service events, which may have conflicts. (They can have conflicts beyond this particular case of multiple children of block `C`, since there may be multiple ERs containing the same service event in different forks, as well as different seals.)

I think the best way to handle all these cases is to allow duplicates directly in the storage layer of `EpochSetups` and `EpochCommits` -- I'll add this and a test case.

Reply:
> Would this be for performance?

Not for performance. Just to note that I was suggesting this:
```
	parent, err := m.blocks.ByID(candidate.Header.ParentID)
	if err != nil {
		return nil, fmt.Errorf("could not get parent (id=%x): %w", block.Header.ParentID, err)
	}
 	// SIXTH: epoch transitions...
 	ops, err := m.handleServiceEvents(parent) 
 	if err != nil { 
 	 	return fmt.Errorf("could not handle service events: %w", err) 
  	} 
```


> when inserting conflicting children of C

I'm not sure if we are talking about the same case. The case I'm talking about is this:

```
A <- B(RA) <- C(SA) <- D
              ^------- E
```

Block C has 2 children: `D` and `E`. When we receive D first, we will call `handleServiceEvents(D)` which store the service events for its parent block (C). And then when we receive E later, we will call `handleServiceEvents(E)` again which will store the service events for its parent block (C) again. And storing the service events will [store the epoch setup event](https://github.com/onflow/flow-go/blob/jordan/4866-handle-service-events/state/protocol/badger/mutator.go#L816), which seems [doesn't allow `SkipDuplication`](https://github.com/onflow/flow-go/blob/jordan/4866-handle-service-events/storage/badger/epoch_setups.go#L22), which means, it will fail, right?

Reply:
I see. We can't _only_ pass in the parent block, because we insert the epoch status for the block being inserted (ie. block `D`), _not_ the parent block.
https://github.com/onflow/flow-go/blob/9c91dc26cded866644c114840d7b7241bb21a446/state/protocol/badger/mutator.go#L858-L859

> Block C has 2 children: D and E

That is the same case I'm thinking of 👍 , added `SkipDuplicate` to the service event storage APIs in https://github.com/onflow/flow-go/pull/399/commits/132e243ab527aa4bb3eccf0377cdcee6001fc408 and test case in https://github.com/onflow/flow-go/pull/399/commits/9e5bdaa0f1f8e76a41c4badbdb10812fd6d6b365.